mcts.py

```diff
@@ -0,0 +1,135 @@
+import dataclasses
+import logging
+import typing
+
+import numpy as np
+import torch
+
+import abstract
+import skyjo_immutable as sj
+import skynet
+
+
+@dataclasses.dataclass(slots=True)
+class MCTSNode:
+    game_state: abstract.AbstractImmutableGameState
+    parent: typing.Self | None
+    children: dict[abstract.AbstractGameAction, typing.Self]
+    number_of_visits: int
+    average_visit_value: float
+    is_expanded: bool = False
+    normalized_policy: skynet.PolicyOutput | None = None
+    model_value: skynet.ValueOutput | None = None
+
+    def expand(self, model: abstract.AbstractModel):
+        assert not self.is_expanded, "Node already expanded"
+        with torch.no_grad():
+            model_output = model.predict(self.game_state.numpy())
+        valid_actions_mask = self.game_state.create_valid_actions_mask().reshape(
+            -1, *sj.ACTION_SHAPE
+        )
+        self.normalized_policy = (
+            model_output.policy_output.mask_invalid_and_renormalize(valid_actions_mask)
+        )
+        self.model_value = model_output.value_output
+
+        for action in self.game_state.valid_actions:
+            self.children[action] = MCTSNode(
+                game_state=self.game_state.next_state(action),
+                parent=self,
+                children={},
+                number_of_visits=0,
+                average_visit_value=0,
+            )
+        self.is_expanded = True
+
+    def select_highest_ucb_child(self, c_puct: float) -> typing.Self:
+        max_ucb = -float("inf")
+        best_child = None
+        for action, child in self.children.items():
+            child_ucb = (
+                child.average_visit_value
+                + c_puct
+                * self.normalized_policy.get_action_probability(action)
+                * np.sqrt(self.number_of_visits)
+                / (1 + child.number_of_visits)
+            )
+            if child_ucb > max_ucb:
+                max_ucb = child_ucb
+                best_child = child
+        return best_child
+
+    def action_probabilities(
+        self, temperature: float
+    ) -> dict[abstract.AbstractGameAction, float]:
+        actions = [action for action, _ in self.children.items()]
+        counts = np.array(
+            [child.number_of_visits for _, child in self.children.items()]
+        )
```

Comment:
probably won't matter in the grand scheme of things but you could consider doing this in a single loop